llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp

```diff
@@ -5891,8 +5910,11 @@ SDValue DAGTypeLegalizer::WidenVecRes_SETCC(SDNode *N) {
     InOp1 = GetWidenedVector(InOp1);
     InOp2 = GetWidenedVector(InOp2);
   } else {
-    InOp1 = DAG.WidenVector(InOp1, SDLoc(N));
-    InOp2 = DAG.WidenVector(InOp2, SDLoc(N));
+    do {
```

Comment:
Type legalization doesn't operate incrementally in individual steps. Newly legalized operations are supposed to be re-legalized later if necessary

Reply:
@arsenm 

For having looked at this with Rob, the problem we are facing here is:
- The legalization strategy looks like this:
  - `v1i16` is marked as `widen` by AArch64 backend and the target type is `v4i16`
  - `v1f16` is marked as `scalarize` and the target type is `f16`
So the legalization strategies don't match and we end up in this block.
Now, the problem is a single call to `WidenVector` will push `v1f16` to `v2f16` and not the expected `v4f16` and the assert line 5900 (of the old code) would break.

There may be a better way to fix that but the newly legalized operations won't get a chance to be re-legalized.

Open to suggestions, admittedly I went with a quick fix here.

FWIW, GISel doesn't have this issue.

Reply:
The widening is only really intended to apply to integer types. A `v1i1 setcc v1f16` having been turned into a `v1i16 setcc v1f16` would be better to scalarize from there if we can tell it that.

Reply:
I think the correct way to handle this is by inserting an explicit conversion instruction from the widened v2f16 to match v4f16. i.e. create a new INSERT_SUBVECTOR v4f16:undef, (v2f16 GetWidenedVector), 0

Reply:
> The widening is only really intended to apply to integer types. A `v1i1 setcc v1f16` having been turned into a `v1i16 setcc v1f16` would be better to scalarize from there if we can tell it that.

Yes, but that's not what the AArch64 lowering wants (see https://github.com/llvm/llvm-project/blob/f0cdf4b468f6ee48b0d0d51ce78145455e2f07a6/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L24550).
I believe the rational is `v1i16` is not natively supported so we widen to the larger supported vector size.
Changing that would fix this particular issue, but I believe it would create worse code in a bunch of cases.

Reply:
> I think the correct way to handle this is by inserting an explicit conversion instruction from the widened v2f16 to match v4f16. i.e. create a new INSERT_SUBVECTOR v4f16:undef, (v2f16 GetWidenedVector), 0

Hmm, technically this is what this loop does, no?

Reply:
Let me re-phrase, we could change the code to explicitly create an insert_subvector here, but I don't see how this is better :).

Reply:
> Yes, but that's not what the AArch64 lowering wants (see https://github.com/llvm/llvm-project/blob/f0cdf4b468f6ee48b0d0d51ce78145455e2f07a6/llvm/lib/Target/AArch64/AArch64ISelLowering.cpp#L24550).
> I believe the rational is v1i16 is not natively supported so we widen to the larger supported vector size.
> Changing that would fix this particular issue, but I believe it would create worse code in a bunch of cases.

Yeah, The point of that code is to act upon integer types, to keep them in vector registers as opposed to scalarizing them to gprs. A `fcmp uno <1 x half>` is really a floating point operation though, even if it's return type becomes a integer when it gets transformed to a `v1i16 setcc v1f16`. As an operation it still makes sense to scalarize it, not widen.

I agree that we shouldn't be changing the way that integer types are handled in general, but if there was a way to specify that a setcc with fp operands should really be treated like fp types, that would avoid a lot of the awkwardness here where we attempt to widen it only to scalarize again.

Reply:
> As an operation it still makes sense to scalarize it, not widen.

I agree, but could you refresh my memory: is it something we have a control on?

IIRC, the selection of a strategy for type legalization is agnostic of the particular instruction being legalized and happens before the custom lowering, but it's been a while since I touched SDISel.

Reply:
Thats always the question, and Im not sure there is a way to specify the types for operations separately. Looking through the example though, it appears that the `v1i16 setcc v1f16` is being created in the AArch64 backend inside performVSelectCombine.

I've put up # as a possible fix. It just tries to prevent the generation of `v1i16 setcc v1f16`, which so long as they dont get generated from anywhere else seems to work OK, even if the code generated it not necessarily optimal. Let me know what you think.

Reply:
Works for me.